### PR TITLE
Add sum-all program

### LINKED ISSUE #3: Add sum-all program
_Instructions will be added in it's Readme_

### PR DESCRIPTION
**Pull request by vaibhavarora14**
**Program strategy -**
- Sorts input array to find `smallNumber` at `0` position and `largerNumber` at `1` position.
- Get a range generator object to fetch numbers between `small` and `large` number of input array.
- Iterate over generator and `add` them to a variable, which is meant to start from 0 and increase as numbers getting added.
- Return value which was holding result from `addition` operation in previous step.

### CODE REVIEW

File: js-exercises/sum-all/sumAll.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+const isNumber = (arg) => typeof arg === 'number';
+
+const areNumbers = (args) => args.every(isNumber);
```
Comment (ArfatSalman) [inline js-exercises/sum-all/sumAll.js:3]: This would be true for `NaN` as well.

File: js-exercises/sum-all/sumAll.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+function* getRange(start, end) {
```
Comment (ArfatSalman) [inline js-exercises/sum-all/sumAll.js]: A for loop is preferred since it is less likely to make an infinite loop there.

Comment (ArfatSalman) [reply js-exercises/sum-all/sumAll.js]: Also, start can have a good default value such as 0.
```suggestion
function* getRange(start = 0, end) {
```

File: js-exercises/sum-all/sumAll.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  let temp = start;
+  while (temp <= end) {
```
Comment (ArfatSalman) [inline js-exercises/sum-all/sumAll.js:15]: It is really important to consider whether whether you want to include the last element as you did here, to not include it. For the record, python's range does not include the last element.

File: js-exercises/sum-all/sumAll.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    yield temp;
+    temp += 1;
+  }
+}
+
+function sumAll(arrayOfNumbers) {
+  let inputArray = arrayOfNumbers;
```
Comment (ArfatSalman) [inline js-exercises/sum-all/sumAll.js]: Why alias `arrayOfNumbers`?

File: js-exercises/sum-all/sumAll.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    throw TypeError('Input array should have numbers only');
+  }
+
+  // program is for numeric integers without decimal, so removing the decimals
+  inputArray = inputArray.map(number => Math.floor(number));
+
+  const sortedArray = inputArray.sort((a, b) => a - b);
+
+  let result = 0;
+  const rangeIterator = getRange(sortedArray[0], sortedArray[1]);
```
Comment (ArfatSalman) [inline js-exercises/sum-all/sumAll.js]:
```suggestion
  const rangeIterator = getRange(lowerBound, upperBound);
```

File: js-exercises/sum-all/sumAll.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+  for (const number of rangeIterator) {
```
Comment (ArfatSalman) [inline js-exercises/sum-all/sumAll.js]:
```suggestion
  for (const number of getRange(lowerBound, upperBound)) {
```

**Review by ArfatSalman — CHANGES_REQUESTED** [review]
Also try the arithmetic mean formula.

File: js-exercises/sum-all/sumAll.js (hunk 1 above)
Comment (vaibhavarora14) [reply js-exercises/sum-all/sumAll.js:3]: right, issue fixed in 1e9a59a7b31f1961f18cfa90dab8749e73646be6

File: js-exercises/sum-all/sumAll.js (hunk 4 above)
Comment (vaibhavarora14) [reply js-exercises/sum-all/sumAll.js]: variable names improved with b093b7825633e962b20ad8258369b28eaa7ee122

File: js-exercises/sum-all/sumAll.js (hunk 5 above)
Comment (vaibhavarora14) [reply js-exercises/sum-all/sumAll.js]: You were looking for something like
21c5e9e924a14adc1d49eca299ae56131e5a938e && a984294784b6516bdf08e2ee64facfecc46069ab ?

File: js-exercises/sum-all/sumAll.js (hunk 3 above)
Comment (vaibhavarora14) [reply js-exercises/sum-all/sumAll.js:15]: yes, right.
current scope of sum-all specs says to return sum of all numbers between `start`, `end`, including them.`getRange` currently works according to that.

If `getRange` has to grow further. it should be entertaining both types of requirement.
- return numbers between `start` and `end`. (Note: Excluding `start` as well unlike Python one).
- return numbers between `start` and `end`, including parameters passed.

What say?

**Comment by vaibhavarora14** [comment]
@ArfatSalman I didn't understood what you want to say with https://github.com/pesto-students/deep-dive-varora1406/pull/4#pullrequestreview-400413577.

Should I make program for mean, as sum-all?
OR did you wanted me to realize about tight coupled code here, which i recently loosen up with https://github.com/pesto-students/deep-dive-varora1406/pull/4/commits/73146eb2376b8b8ccaea6604211f76fcd31ef12b

**Pull request closed without merging**